tket/src/Transformations/ControlledGates.cpp

```diff
@@ -432,22 +432,27 @@ static Circuit lemma71(
   if (rep.n_gates() != correct_gate_count)
     throw ControlDecompError("Error in Lemma 7.1: Gate count is incorrect");
   auto [vit, vend] = boost::vertices(rep.dag);
+  VertexSet bin;
   for (auto next = vit; vit != vend; vit = next) {
     ++next;
     Vertex v = *vit;
-    if (rep.get_OpType_from_Vertex(v) == OpType::CRy) {
+    if (!bin.contains(v) && rep.get_OpType_from_Vertex(v) == OpType::CRy) {
       Expr v_angle = rep.get_Op_ptr_from_Vertex(v)->get_params()[0];
       Circuit cry_replacement = CircPool::CRy_using_CX(v_angle);
       Subcircuit sub{rep.get_in_edges(v), rep.get_all_out_edges(v), {v}};
-      rep.substitute(cry_replacement, sub, Circuit::VertexDeletion::Yes);
+      rep.substitute(cry_replacement, sub, Circuit::VertexDeletion::No);
+      bin.insert(v);
     }
-    if (rep.get_OpType_from_Vertex(v) == OpType::CU1) {
+    if (!bin.contains(v) && rep.get_OpType_from_Vertex(v) == OpType::CU1) {
```

Comment:
Would changing this to `else if` solve the pointer referencing problem?

Reply:
Yes the code could definitely be better structured...

Reply:
Done.